File: leetcode/JavaScript实现/100-199题/0145_二叉树的后序遍历.js
// 1、递归
var postorderTraversal = function(root) {
    let ans = [];

    let postOrder = (temp, arr) => {
        if (temp === null)
            return;
        postOrder(temp.left, arr);
        postOrder(temp.right, arr);
        ans.push(temp.val);
    };

    postOrder(root, []);

    return ans;
};

// 2、迭代 非递归实现
var postorderTraversal = function(root) {
    let stack = [];
    let ans = [];

    if (!root)
        return [];
    stack.push(root);

    while (stack.length !== 0) {
        let temp = stack.pop();
        ans.push(temp.val);
        if (temp.left !== null) 
            stack.push(temp.left);
        if (temp.right !== null) 
            stack.push(temp.right);
    }

    return ans.reverse();
};

File: leetcode/C++实现/100-199题/0145_二叉树的后序遍历.cpp
// 自己习惯的写法，也是王道书上的代码
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<TreeNode*> s;
        TreeNode *p = root;
        TreeNode *r = NULL;
        
        while (p || !s.empty()) {
            if (p) {
                s.push(p);
                p = p->left;
            } else {
                TreeNode *temp = s.top();
                
                if (temp->right && temp->right != r) {
                    p = temp->right;
                    s.push(p);
                    p = p->left;
                } else {
                    s.pop();
                    ans.push_back(temp->val);
                    r = temp;
                    p = NULL;
                }
            }
        }

        return ans;
    }
};